File: clients/android/app/src/main/java/com/familyconnect/app/ControlProtocol.java
package com.familyconnect.app;

import com.google.gson.*;
import java.nio.charset.StandardCharsets;
import java.security.*;
import java.util.*;
import javax.crypto.*;
import javax.crypto.spec.*;
import org.bouncycastle.crypto.params.*;
import org.bouncycastle.crypto.signers.Ed25519Signer;
import org.bouncycastle.crypto.generators.HKDFBytesGenerator;
import org.bouncycastle.crypto.digests.SHA256Digest;
import static com.familyconnect.app.ControlJson.*;

/** Verification only: no carrier, storage, revision acceptance, or VPN mutations. */
final class ControlProtocol {
    static final String CONFIG="family-connect/control-config/v1", ACK="family-connect/control-ack/v1";
    static final class Rejected extends Exception {
        final String category;Rejected(String category){super(category);this.category=category;}
    }
    static final class Verified {
        private final JsonObject state;final String digest;
        Verified(JsonObject state,String digest){this.state=state.deepCopy();this.digest=digest;}
        JsonObject state(){return state.deepCopy();}
    }
    static byte[] base64(String value,boolean canonical) {
        require(value.matches("[A-Za-z0-9+/]*={0,2}") && value.length()%4==0);
        byte[] bytes=Base64.getDecoder().decode(value);
        if(canonical)require(Base64.getEncoder().encodeToString(bytes).equals(value));return bytes;
    }
    static byte[] sha(byte[] value)throws GeneralSecurityException{return MessageDigest.getInstance("SHA-256").digest(value);}
    static String hash(byte[] value)throws GeneralSecurityException {
        StringBuilder result=new StringBuilder();for(byte b:sha(value))result.append(String.format(Locale.ROOT,"%02x",b&255));return result.toString();
    }
    private static boolean signature(byte[] key,byte[] sig,String purpose,byte[] value) {
        if(key.length!=32 || sig.length!=64)return false;
        Ed25519Signer verifier=new Ed25519Signer();verifier.init(false,new Ed25519PublicKeyParameters(key,0));
        byte[] domain=(purpose+"\0").getBytes(StandardCharsets.UTF_8);
        verifier.update(domain,0,domain.length);verifier.update(value,0,value.length);return verifier.verifySignature(sig);
    }
    static JsonObject verifyAck(byte[] raw)throws Rejected {
        try {
            require(raw.length<=4096);JsonElement outerValue=parse(raw);fields(outerValue,"body signature");JsonObject outer=outerValue.getAsJsonObject();
            JsonElement bodyValue=outer.get("body");fields(bodyValue,"schema_version device public_identity envelope_hash config_id sequence status error timestamp ack_id");JsonObject body=bodyValue.getAsJsonObject();
            require(integer(body.get("schema_version"),0)==1);integer(body.get("sequence"),0);integer(body.get("timestamp"),0);
            require(Arrays.asList("RECEIVED REJECTED APPLIED COMMITTED ROLLED_BACK FAILED".split(" ")).contains(text(body.get("status"))));
            require(Arrays.asList("NONE SIZE CLOCK SIGNATURE MALFORMED SCHEMA TARGET SIGNER STRUCTURE LEASE CLIENT_VERSION UNSUPPORTED_TRANSPORT_VERSION REPLAY PREVIOUS_HASH APPLY HEALTH RECOVERY ROLLBACK".split(" ")).contains(text(body.get("error"))));
            require(text(body.get("envelope_hash")).matches("[0-9a-f]{64}") && text(body.get("ack_id")).matches("[0-9a-f]{64}"));
            require(body.get("config_id").isJsonNull() || text(body.get("config_id")).matches("[a-zA-Z0-9_-]{1,64}"));
            byte[] key=base64(text(body.get("public_identity")),true);require(key.length==64);
            require(hash(key).substring(0,32).equals(text(body.get("device"))));
            require(signature(Arrays.copyOfRange(key,32,64),base64(text(outer.get("signature")),false),ACK,canonical(body,false)));
            require(hash(canonical(body,true)).equals(text(body.get("ack_id"))));return body.deepCopy();
        } catch(Exception e){throw new Rejected("REJECT");}
    }
    private static byte[] decrypt(byte[] cipher,byte[] privateKey,byte[] publicIdentity)throws Exception {
        require(privateKey.length==64 && publicIdentity.length==64 && cipher.length>=96);
        byte[] shared=new byte[32],derived=new byte[64];
        try {
            new X25519PrivateKeyParameters(privateKey,0).generateSecret(new X25519PublicKeyParameters(cipher,0),shared,0);
            HKDFBytesGenerator hkdf=new HKDFBytesGenerator(new SHA256Digest());
            hkdf.init(new HKDFParameters(shared,Arrays.copyOf(sha(publicIdentity),16),null));hkdf.generateBytes(derived,0,64);
            Mac mac=Mac.getInstance("HmacSHA256");mac.init(new SecretKeySpec(Arrays.copyOf(derived,32),"HmacSHA256"));
            mac.update(cipher,32,cipher.length-64);require(MessageDigest.isEqual(mac.doFinal(),Arrays.copyOfRange(cipher,cipher.length-32,cipher.length)));
            Cipher aes=Cipher.getInstance("AES/CBC/PKCS5Padding");
            aes.init(Cipher.DECRYPT_MODE,new SecretKeySpec(Arrays.copyOfRange(derived,32,64),"AES"),new IvParameterSpec(cipher,32,16));
            return aes.doFinal(cipher,48,cipher.length-80);
        } finally {Arrays.fill(shared,(byte)0);Arrays.fill(derived,(byte)0);}
    }
    static int[] version(String value) {
        require(value.matches("(0|[1-9][0-9]{0,5})\\.(0|[1-9][0-9]{0,5})\\.(0|[1-9][0-9]{0,5})"));
        String[] parts=value.split("\\.");return new int[]{Integer.parseInt(parts[0]),Integer.parseInt(parts[1]),Integer.parseInt(parts[2])};
    }
    private static boolean equalsString(JsonElement value,String expected){return value!=null && value.isJsonPrimitive() && value.getAsJsonPrimitive().isString() && value.getAsString().equals(expected);}
    static Verified verifyConfiguration(byte[] raw,byte[] anchor,byte[] rnsPrivate,byte[] publicIdentity,String wgPublic,String clientVersion,long now)throws Rejected {
        if(now<0)throw new Rejected("CLOCK");if(raw==null || raw.length>65536)throw new Rejected("SIZE");byte[] cipher;
        try {
            JsonElement value=parse(raw);fields(value,"ciphertext signature");JsonObject outer=value.getAsJsonObject();cipher=base64(text(outer.get("ciphertext")),false);
            if(!signature(anchor,base64(text(outer.get("signature")),false),CONFIG,cipher))throw new Rejected("SIGNATURE");
        }catch(Rejected e){throw e;}catch(Exception e){throw new Rejected("MALFORMED");}
        JsonObject state;
        try {
            byte[] plain=decrypt(cipher,rnsPrivate,publicIdentity);JsonElement value;
            try{value=parse(plain);}finally{Arrays.fill(plain,(byte)0);}
            require(value.isJsonObject());state=value.getAsJsonObject();JsonElement schema=state.get("schema_version");
            if(schema==null || !schema.isJsonPrimitive() || !schema.getAsJsonPrimitive().isNumber() || !schema.getAsString().equals("2"))throw new Rejected("SCHEMA");
            if(!equalsString(state.get("audience"),CONFIG) || !equalsString(state.get("recipient"),hash(publicIdentity).substring(0,32)))throw new Rejected("TARGET");
            if(!equalsString(state.get("signer_key_id"),hash(anchor)))throw new Rejected("SIGNER");
            JsonElement profiles=state.get("transport_profiles");
            if(profiles!=null && profiles.isJsonArray())for(JsonElement p:profiles.getAsJsonArray())
                if(p.isJsonObject() && equalsString(p.getAsJsonObject().get("transport"),"amneziawg") && equalsString(p.getAsJsonObject().get("transport_version"),"3.1"))throw new Rejected("UNSUPPORTED_TRANSPORT_VERSION");
            ControlProfiles.validate(state);
        }catch(Rejected e){throw e;}catch(Exception e){throw new Rejected("STRUCTURE");}
        try {
            if(!text(state.get("wireguard_public_key")).equals(wgPublic))throw new Rejected("TARGET");
            if(now<integer(state.get("issued_at"),1) || now>=integer(state.get("expires_at"),1))throw new Rejected("LEASE");
            int[] minimum=version(text(state.get("min_client_version"))),client=version(clientVersion);
            for(int n=0;n<3;n++){if(minimum[n]>client[n])throw new Rejected("CLIENT_VERSION");if(minimum[n]<client[n])break;}
            return new Verified(state,hash(raw));
        }catch(Rejected e){throw e;}catch(Exception e){throw new Rejected("STRUCTURE");}
    }
}

File: clients/windows/Core/ControlProtocol.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;

namespace FamilyConnect;

public sealed class ControlRejected(string category) : FormatException(category)
{
    public string Category { get; } = category;
}

// Carrier-independent verification only. This class grants no application authority:
// revision/chain acceptance and durable recovery belong to the future journal.
public static class ControlProtocol
{
    internal const string ConfigPurpose = "family-connect/control-config/v1";
    internal const string AckPurpose = "family-connect/control-ack/v1";
    internal static void Require(bool condition) { if (!condition) throw new FormatException("invalid control data"); }
    internal static bool Matches(string text, string pattern) => Regex.IsMatch(text, "\\A(?:"+pattern+")\\z", RegexOptions.CultureInvariant);
    internal static string Text(JsonElement value) => value.GetString() ?? throw new FormatException("missing string");
    internal static long Integer(JsonElement value, long minimum = 0)
    {
        Require(value.ValueKind == JsonValueKind.Number && Matches(value.GetRawText(), "-?[0-9]+"));
        Require(value.TryGetInt64(out var number) && number >= minimum); return number;
    }
    internal static void Fields(JsonElement value, string fields)
    {
        Require(value.ValueKind == JsonValueKind.Object);
        Require(value.EnumerateObject().Select(p => p.Name).ToHashSet(StringComparer.Ordinal).SetEquals(fields.Split(' ')));
    }
    internal static JsonElement Parse(byte[] raw)
    {
        // Reject duplicates recursively, including escaped spellings of the same key.
        using var doc = JsonDocument.Parse(raw, new JsonDocumentOptions { MaxDepth = 64 });
        void Visit(JsonElement value)
        {
            if (value.ValueKind == JsonValueKind.Object)
            {
                var names = new HashSet<string>(StringComparer.Ordinal);
                foreach (var p in value.EnumerateObject()) { Require(names.Add(p.Name)); Visit(p.Value); }
            }
            else if (value.ValueKind == JsonValueKind.Array) foreach (var item in value.EnumerateArray()) Visit(item);
        }
        Visit(doc.RootElement); return doc.RootElement.Clone();
    }
    internal static byte[] Base64(string text, bool canonical = false)
    {
        Require(Matches(text, "[A-Za-z0-9+/]*={0,2}"));
        var bytes = Convert.FromBase64String(text);
        if (canonical) Require(Convert.ToBase64String(bytes) == text);
        return bytes;
    }
    internal static string Hash(byte[] bytes) => Convert.ToHexStringLower(SHA256.HashData(bytes));
    internal static bool Signature(byte[] key, byte[] signature, string purpose, byte[] body)
    {
        if (key.Length != 32 || signature.Length != 64) return false;
        var signer = new Ed25519Signer(); signer.Init(false, new Ed25519PublicKeyParameters(key, 0));
        var domain = Encoding.UTF8.GetBytes(purpose+"\0");
        signer.BlockUpdate(domain, 0, domain.Length); signer.BlockUpdate(body, 0, body.Length);
        return signer.VerifySignature(signature);
    }
    // ACK strings are constrained to ASCII below. Python json.dumps escapes neither
    // '+' nor '/', whereas the default .NET JSON encoder does escape '+'.
    internal static byte[] AckCanonical(JsonElement body, bool omitId = false)
    {
        using var stream = new MemoryStream();
        using (var writer = new Utf8JsonWriter(stream, new JsonWriterOptions { Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping }))
        {
            writer.WriteStartObject();
            foreach (var p in body.EnumerateObject().OrderBy(p => p.Name, StringComparer.Ordinal))
            {
                if (omitId && p.Name == "ack_id") continue;
                writer.WritePropertyName(p.Name);
                if (p.Value.ValueKind == JsonValueKind.Number) writer.WriteNumberValue(Integer(p.Value));
                else p.Value.WriteTo(writer);
            }
            writer.WriteEndObject();
        }
        return stream.ToArray();
    }
    public static JsonElement VerifyAck(byte[] raw)
    {
        try
        {
            Require(raw.Length <= 4096); var outer = Parse(raw); Fields(outer, "body signature");
            var body = outer.GetProperty("body");
            Fields(body, "schema_version device public_identity envelope_hash config_id sequence status error timestamp ack_id");
            Require(Integer(body.GetProperty("schema_version")) == 1);
            Integer(body.GetProperty("sequence")); Integer(body.GetProperty("timestamp"));
            Require("RECEIVED REJECTED APPLIED COMMITTED ROLLED_BACK FAILED".Split(' ').Contains(Text(body.GetProperty("status"))));
            Require("NONE SIZE CLOCK SIGNATURE MALFORMED SCHEMA TARGET SIGNER STRUCTURE LEASE CLIENT_VERSION UNSUPPORTED_TRANSPORT_VERSION REPLAY PREVIOUS_HASH APPLY HEALTH RECOVERY ROLLBACK".Split(' ').Contains(Text(body.GetProperty("error"))));
            Require(Matches(Text(body.GetProperty("envelope_hash")), "[0-9a-f]{64}"));
            Require(Matches(Text(body.GetProperty("ack_id")), "[0-9a-f]{64}"));
            var id = body.GetProperty("config_id");
            Require(id.ValueKind == JsonValueKind.Null || Matches(Text(id), "[a-zA-Z0-9_-]{1,64}"));
            var key = Base64(Text(body.GetProperty("public_identity")), true); Require(key.Length == 64);
            Require(Hash(key)[..32] == Text(body.GetProperty("device")));
            Require(Signature(key[32..], Base64(Text(outer.GetProperty("signature"))), AckPurpose, AckCanonical(body)));
            Require(Hash(AckCanonical(body, true)) == Text(body.GetProperty("ack_id")));
            return body.Clone();
        }
        catch (Exception e) when (Invalid(e)) { throw new ControlRejected("REJECT"); }
    }
    internal static bool Invalid(Exception e) => e is FormatException or JsonException or InvalidOperationException or ArgumentException or KeyNotFoundException or CryptographicException or Org.BouncyCastle.Crypto.CryptoException or OverflowException;
    static byte[] Decrypt(byte[] cipher, byte[] privateKey, byte[] publicIdentity)
    {
        Require(privateKey.Length == 64 && publicIdentity.Length == 64 && cipher.Length >= 96);
        var shared = new byte[32];
        new X25519PrivateKeyParameters(privateKey, 0).GenerateSecret(new X25519PublicKeyParameters(cipher, 0), shared, 0);
        Require(shared.Any(b => b != 0));
        var derived = HKDF.DeriveKey(HashAlgorithmName.SHA256, shared, 64, SHA256.HashData(publicIdentity)[..16]);
        try
        {
            var token = cipher[32..];
            Require(CryptographicOperations.FixedTimeEquals(HMACSHA256.HashData(derived[..32], token[..^32]), token[^32..]));
            using var aes = Aes.Create(); aes.Key = derived[32..];
            return aes.DecryptCbc(token[16..^32], token[..16], PaddingMode.PKCS7);
        }
        finally { CryptographicOperations.ZeroMemory(shared); CryptographicOperations.ZeroMemory(derived); }
    }
    internal static Version VersionValue(string text)
    {
        Require(Matches(text, "(0|[1-9][0-9]{0,5})\\.(0|[1-9][0-9]{0,5})\\.(0|[1-9][0-9]{0,5})"));
        return Version.Parse(text);
    }
    public sealed record VerifiedConfiguration(JsonElement State, string Digest);
    public static VerifiedConfiguration VerifyConfiguration(byte[] raw, byte[] anchor, byte[] rnsPrivate,
        byte[] publicIdentity, string wireguardPublicKey, string clientVersion, long now)
    {
        if (now < 0) throw new ControlRejected("CLOCK");
        if (raw.Length > 65536) throw new ControlRejected("SIZE");
        byte[] cipher;
        try
        {
            var outer = Parse(raw); Fields(outer, "ciphertext signature");
            cipher = Base64(Text(outer.GetProperty("ciphertext")));
            if (!Signature(anchor, Base64(Text(outer.GetProperty("signature"))), ConfigPurpose, cipher)) throw new ControlRejected("SIGNATURE");
        }
        catch (ControlRejected) { throw; }
        catch (Exception e) when (Invalid(e)) { throw new ControlRejected("MALFORMED"); }
        JsonElement state;
        try
        {
            var plain = Decrypt(cipher, rnsPrivate, publicIdentity);
            try { state = Parse(plain); } finally { CryptographicOperations.ZeroMemory(plain); }
            Require(state.ValueKind == JsonValueKind.Object);
            if (!state.TryGetProperty("schema_version", out var schema) || schema.ValueKind != JsonValueKind.Number || schema.GetRawText() != "2") throw new ControlRejected("SCHEMA");
            if (!state.TryGetProperty("audience", out var audience) || audience.ValueKind != JsonValueKind.String || Text(audience) != ConfigPurpose ||
                !state.TryGetProperty("recipient", out var recipient) || recipient.ValueKind != JsonValueKind.String || Text(recipient) != Hash(publicIdentity)[..32]) throw new ControlRejected("TARGET");
            if (!state.TryGetProperty("signer_key_id", out var signer) || signer.ValueKind != JsonValueKind.String || Text(signer) != Hash(anchor)) throw new ControlRejected("SIGNER");
            if (state.TryGetProperty("transport_profiles", out var profiles) && profiles.ValueKind == JsonValueKind.Array)
                foreach (var p in profiles.EnumerateArray())
                    if (p.ValueKind == JsonValueKind.Object && p.TryGetProperty("transport", out var t) && t.ValueKind == JsonValueKind.String && Text(t) == "amneziawg" && p.TryGetProperty("transport_version", out var v) && v.ValueKind == JsonValueKind.String && Text(v) == "3.1") throw new ControlRejected("UNSUPPORTED_TRANSPORT_VERSION");
            ControlProfiles.Validate(state);
        }
        catch (ControlRejected) { throw; }
        catch (Exception e) when (Invalid(e)) { throw new ControlRejected("STRUCTURE"); }
        if (Text(state.GetProperty("wireguard_public_key")) != wireguardPublicKey) throw new ControlRejected("TARGET");
        if (now < Integer(state.GetProperty("issued_at")) || now >= Integer(state.GetProperty("expires_at"))) throw new ControlRejected("LEASE");
        if (VersionValue(Text(state.GetProperty("min_client_version"))) > VersionValue(clientVersion)) throw new ControlRejected("CLIENT_VERSION");
        return new(state, Hash(raw));
    }
}
